src/parser/parser.c

```diff
@@ -46,107 +84,108 @@ void parse_line(md_node *root, const char *line, int line_length) {
 
-    set_header_data(new_child_node, line, line_length);
-    append_to_root(root, new_child_node);
+  switch (current_line_type) {
+  case LINE_EMPTY:
+    if (root->last_child != NULL && root->last_child->type == NODE_CODE_BLOCK &&
+        root->last_child->user_data != MODE_PROCESSED) {
 
-  } else if (is_codeblock_indicator(line, line_length) && strcmp(line, "```mermaid")) {
-    switch (root->user_data) {
-    case MODE_CODE:
-      root->user_data = MODE_STARTNEW;
-      break;
-    case MODE_EMPTY:
-    case MODE_STARTNEW:
-      root->user_data = MODE_CODE;
+      // we don't want to process last and add the code data
+      set_code_data(root->last_child, line, line_length);
+
+    } else {
+      process_last_node(root->last_child);
+    }
+    break;
+
+  case LINE_CODE_DELIM:
+    // check if previous one was code block
+    if (root->last_child != NULL && root->last_child->type == NODE_CODE_BLOCK &&
+        root->last_child->user_data != MODE_PROCESSED) {
+      root->last_child->user_data = MODE_PROCESSED;
+
+      printf("end code block\n");
+    } else {
+      // start new node
+      process_last_node(root->last_child);
+
+      printf("start code block\n");
       new_child_node = create_empty_md_node(NODE_CODE_BLOCK);
       set_code_language(new_child_node, line, line_length);
       append_to_root(root, new_child_node);
+    }
+    break;
+
+  case LINE_LISTITEM:
+
+    if (root->last_child == NULL || (root->last_child->type != NODE_LIST) ||
+        (root->last_child->user_data == MODE_PROCESSED)) {
+
+      new_child_node = create_empty_md_node(NODE_LIST);
+      append_to_root(root, new_child_node);
+    }
+
+    switch (root->last_child->type) {
+    case NODE_LIST:
+      set_item_data(root->last_child, line, line_length);
       break;
-    case MODE_APPENDPARA:
-      printf("why here\n");
-      abort();
     default:
-      printf("Unidentified mode to start code block %d\n", root->user_data);
-      abort();
+      printf("should not reach\n");
+      break;
     }
-  } else if (strcmp(line, "```mermaid") == 0) {
-    printf("entering mermaid block\n");
-    in_mermaid_block = 1;
-    free(mermaid_code); 
-    mermaid_code = calloc(1, sizeof(char));
-  } else if (in_mermaid_block) {
-      // printf("Debug line: [%.*s], length: %d\n", line_length, line, line_length);
-      // printf("CHECK: ");
-      // printf("%d", strncmp(line, "```\r\n", 3));
-      // printf("\n");
-      // It's really weird why it's not 0 and 13, wasted a lot of time here
-      // Fix it please if you can
-      if (strncmp(line, "```", 3) == 13) {
-          // Exiting the Mermaid block
-          printf("exiting mermaid block\n");
-          in_mermaid_block = 0;
-          printf("%s", mermaid_code);
-          printf("\n");
-          md_node *mermaid_node = create_md_node(NODE_MERMAID_DIAGRAM, mermaid_code, strlen(mermaid_code), NULL, NULL, root, NULL, NULL, MODE_NONE);
```

Comment:
where is the `NODE_MERMAID_DIAGRAM` node created if this is deleted? i might have overlooked but i couldnt find it

Reply:
changed it so mermaid code is in node_code_block